### PR TITLE
[Datastore] Support alibaba OSS

### PR DESCRIPTION
**Pull request by yakovga**

### CODE REVIEW

File: tests/integration/alibaba_oss/test-alibaba-oss.yml [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+env:
+  # Bucket name
+  bucket_name:
+  # Access key parameters as created by ALIBABA
+  ALIBABA_ACCESS_KEY_ID:
+  ALIBABA_SECRET_ACCESS_KEY:
+  ALIBABA_ENDPOINT_URL:
```
Comment (liranbg) [inline tests/integration/alibaba_oss/test-alibaba-oss.yml]: missing newline

File: tests/integration/alibaba_oss/test_alibaba_oss.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+    config = yaml.safe_load(fp)
+
+test_filename = here / "test.txt"
+with open(test_filename) as f:
+    test_string = f.read()
+
+# Used to test dataframe functionality (will be saved as csv)
+test_df_string = "col1,col2,col3\n1,2,3"
+
+credential_params = ["ALIBABA_ACCESS_KEY_ID", "ALIBABA_SECRET_ACCESS_KEY","ALIBABA_ENDPOINT_URL"]
```
Comment (liranbg) [inline tests/integration/alibaba_oss/test_alibaba_oss.py]: suggest to move to test class properties, keeping file declared variables not exposed if not needed and keeping file code more concise

File: mlrun/datastore/alibaba_oss.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+
+    def get_bucket_and_key(self, key):
+        path = self._join(key)[1:]
+        return self.endpoint, path
+
+    def upload(self, key, src_path):
+        bucket, key = self.get_bucket_and_key(key)
+        oss = oss2.Bucket(self.auth, self.endpoint_url, bucket)
+        oss.put_object(key, open(src_path, "rb"))
```
Comment (liranbg) [inline mlrun/datastore/alibaba_oss.py:78]: while not exposed or supported yet, can you add the delete as well? would be used later on during mlrun 1.7/1.8 :)

File: tests/integration/alibaba_oss/test_alibaba_oss.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    for param in needed_params:
+        if not env_params.get(param):
+            return False
+    return True
+
+
+@pytest.mark.skipif(not alibaba_oss_configured(), reason="ALIBABA OSS parameters not configured")
+@pytest.mark.parametrize("use_datastore_profile", [False])
+class TestALIBABAOSS:
+    def _make_target_names(
```
Comment (liranbg) [inline tests/integration/alibaba_oss/test_alibaba_oss.py]: move private functions down, keep public up - make it more readable

**Review by liranbg — COMMENTED** [review]
Hi @yakovga ! Thank you for this contribution, I have left minor comments
Let me know if you have further questions

File: mlrun/datastore/alibaba_oss.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+class OSSStore(DataStore):
+    using_bucket = True
+
+    def __init__(self, parent, schema, name, endpoint="", secrets: dict = None):
+        super().__init__(parent, name, schema, endpoint, secrets)
+        # will be used in case user asks to assume a role and work through fsspec
+        self._temp_credentials = None
+
+        self.headers = None
```
Comment (theSaarco) [inline mlrun/datastore/alibaba_oss.py]: Is this a copy-paste leftover? These are not used anywhere.

File: mlrun/datastore/alibaba_oss.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+
+        access_key_id = self._get_secret_or_env("ALIBABA_ACCESS_KEY_ID")
+        secret_key = self._get_secret_or_env("ALIBABA_SECRET_ACCESS_KEY")
+        endpoint_url = self._get_secret_or_env("ALIBABA_ENDPOINT_URL")
+        if access_key_id or secret_key:
```
Comment (theSaarco) [inline mlrun/datastore/alibaba_oss.py]: What if none are provided? I'm not very much familiar with `oss`, but does it have an anonymous mode where you don't need credentials? If not, then I would explode if those are not given.

File: mlrun/datastore/alibaba_oss.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+            self.auth = oss2.Auth(access_key_id, secret_key)
+            self.endpoint_url = endpoint_url
```
Comment (theSaarco) [inline mlrun/datastore/alibaba_oss.py:40]: Should this be in the same `if` block as the creds? If there is an anonymous mode, is it possible that you'd want to still provide an endpoint even though no creds are needed?

**Review by theSaarco — CHANGES_REQUESTED** [review]
@yakovga - welcome aboard! Overall looks pretty good. I have some minor comments.
Can you please also add a section to https://github.com/mlrun/mlrun/blob/development/docs/store/datastore.md documenting the credential parameters for this store?

File: tests/integration/alibaba_oss/test_alibaba_oss.py [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    env_params = config["env"]
+    needed_params = ["bucket_name", *credential_params, *extra_params]
+    for param in needed_params:
+        if not env_params.get(param):
+            return False
+    return True
+
+
+@pytest.mark.skipif(not alibaba_oss_configured(), reason="ALIBABA OSS parameters not configured")
+@pytest.mark.parametrize("use_datastore_profile", [False])
```
Comment (assaf758) [inline tests/integration/alibaba_oss/test_alibaba_oss.py:62]: Can I assume datastore-profiles will come in follow-up PR? We recently completed full coverage of ds-profiles across all datastores..

File: tests/integration/alibaba_oss/test_alibaba_oss.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+class TestALIBABAOSS:
```
Comment (gtopper) [inline tests/integration/alibaba_oss/test_alibaba_oss.py]:
```suggestion
class TestAlibabaOssDataStore:
```
would be in line with equivalent suites e.g. `TestHttpDataStore` and `TestV3ioDataStore`.

File: tests/integration/alibaba_oss/test_alibaba_oss.py [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        self._secret_access_key = config["env"].get("ALIBABA_SECRET_ACCESS_KEY")
+
+        object_dir = "test_mlrun_oss_objects"
+        object_file = f"file_{random.randint(0, 1000)}.txt"
+        csv_file = f"file_{random.randint(0,1000)}.csv"
+
+        self.oss = {}
+        self.oss["oss"] = self._make_target_names(
+            "oss://", self._bucket_name, object_dir, object_file, csv_file
+        )
```
Comment (gtopper) [inline tests/integration/alibaba_oss/test_alibaba_oss.py]: Minor style suggestion:
```suggestion
        self.oss = {
            "oss": self._make_target_names(
                "oss://", self._bucket_name, object_dir, object_file, csv_file
            )
        }
```

File: tests/integration/alibaba_oss/test_alibaba_oss.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        return res
+
+    def setup_method(self, method):
+        self._bucket_name = config["env"].get("bucket_name")
+        self._access_key_id = config["env"].get("ALIBABA_ACCESS_KEY_ID")
+        self._secret_access_key = config["env"].get("ALIBABA_SECRET_ACCESS_KEY")
+
+        object_dir = "test_mlrun_oss_objects"
+        object_file = f"file_{random.randint(0, 1000)}.txt"
+        csv_file = f"file_{random.randint(0,1000)}.csv"
```
Comment (gtopper) [inline tests/integration/alibaba_oss/test_alibaba_oss.py]:
```suggestion
        uid = uuid.uuid4()
        object_file = f"file_{uid}.txt"
        csv_file = f"file_{uid}.csv"
```

File: tests/integration/alibaba_oss/test_alibaba_oss.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        # Use "naked" env variables, useful in client-side sdk.
+        for param in credential_params:
+            os.environ[param] = config["env"][param]
+            os.environ.pop(SecretsStore.k8s_env_variable_name_for_secret(param), None)
+
+        self._perform_alibaba_oss_tests(use_datastore_profile)
+
+        # cleanup
+        for param in credential_params:
+            os.environ.pop(param)
```
Comment (gtopper) [inline tests/integration/alibaba_oss/test_alibaba_oss.py]: I know this is copied from another suite, but it seems to me like it should be inside a finally clause.

File: tests/integration/alibaba_oss/test_alibaba_oss.py (hunk 2 above)
Comment (yakovga) [reply tests/integration/alibaba_oss/test_alibaba_oss.py]: the first use of the credential_params is a pytest decorator on the test class itself. we cant use a class attribute outside. I think its the same reason this implementation is present in the other storage test files

File: tests/integration/alibaba_oss/test-alibaba-oss.yml (hunk 1 above)
Comment (yakovga) [reply tests/integration/alibaba_oss/test-alibaba-oss.yml]: added

File: mlrun/datastore/alibaba_oss.py (hunk 3 above)
Comment (yakovga) [reply mlrun/datastore/alibaba_oss.py:78]: added delete

File: tests/integration/alibaba_oss/test_alibaba_oss.py (hunk 4 above)
Comment (yakovga) [reply tests/integration/alibaba_oss/test_alibaba_oss.py]: moved private functions below

File: mlrun/datastore/alibaba_oss.py (hunk 5 above)
Comment (yakovga) [reply mlrun/datastore/alibaba_oss.py]: correct, removed

File: mlrun/datastore/alibaba_oss.py (hunk 6 above)
Comment (yakovga) [reply mlrun/datastore/alibaba_oss.py]: add exception if required values are missing

File: mlrun/datastore/alibaba_oss.py (hunk 7 above)
Comment (yakovga) [reply mlrun/datastore/alibaba_oss.py:40]: add exception if required values are missing

File: tests/integration/alibaba_oss/test_alibaba_oss.py (hunk 8 above)
Comment (yakovga) [reply tests/integration/alibaba_oss/test_alibaba_oss.py:62]: yes, will be added

File: tests/integration/alibaba_oss/test_alibaba_oss.py (hunk 12 above)
Comment (yakovga) [reply tests/integration/alibaba_oss/test_alibaba_oss.py]: moved to finally

**Review by theSaarco — APPROVED** [review]
LGTM.

**Review by liranbg — APPROVED** [review]

**Comment by assaf758** [comment]
please fix lint errors

**Review by assaf758 — APPROVED** [review]
Nice work!
Please add datastore-profile on a follow-up PR (and fix the Copyright on the new files)

**Pull request merged**